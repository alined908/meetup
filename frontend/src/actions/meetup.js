import {GET_MEETUPS_REQUEST, GET_MEETUPS_SUCCESS, GET_MEETUPS_ERROR, 
    ADD_MEETUP, ADD_MEETUP_MEMBER, ADD_EVENT_OPTION, ADD_GLOBAL_MESSAGE, DELETE_MEETUP, EDIT_MEETUP,
    VOTE_MEETUP_EVENT, GET_MEETUP_EVENTS, DELETE_MEETUP_EVENT, ADD_MEETUP_EVENT, EDIT_MEETUP_EVENT} from "../constants/action-types";
import {axiosClient} from '../accounts/axiosClient';
import {history} from '../components/MeetupApp'

export const getMeetups = (data) => async dispatch => {
    dispatch({type: GET_MEETUPS_REQUEST})
    var params;
    if (data.type === "private"){
        params = {type: data.type, ...data.categories && {categories: data.categories}}
    } else {
        params = {
            type: data.type, ...data.categories && {categories: data.categories}, 
            latitude: data.coords.latitude, longitude: data.coords.longitude, 
            radius: data.coords.radius
        }
    }
    
    try {
        const response = await axiosClient.request({
            method: "GET",
            url: "/api/meetups/", 
            headers: {
                "Authorization": `Bearer ${localStorage.getItem('token')}`
            },
            params: params
        })
        setTimeout(() => dispatch({type: GET_MEETUPS_SUCCESS, payload: response.data.meetups}), 100)
    } catch(e){
        console.log(e)
        dispatch({type: GET_MEETUPS_ERROR, payload: e})
    }
}

export const getMeetup = (uri) => async dispatch => {
    try {
        const response = await axiosClient.get(
            `/api/meetups/${uri}/`, {headers: {
                "Authorization": `Bearer ${localStorage.getItem('token')}`
            }}
        )
        // console.log(response)
        dispatch({type: ADD_MEETUP, payload: response.data})
    } catch(e){
        history.push('/404')
    }
}

export const addMeetup = (formProps, redirectOnSuccess) => async dispatch => {
    var params = {...formProps, 
        date: formProps.date.getFullYear()+ "-" + ("0"+(formProps.date.getMonth()+1)).slice(-2) + "-" + ("0" + formProps.date.getDate()).slice(-2)
    }
    try {
        const response = await axiosClient.post(
            `/api/meetups/`, params, {headers: {
                "Authorization": `Bearer ${localStorage.getItem('token')}`
        }})
        dispatch({type: ADD_MEETUP, payload: response.data.meetup})
        console.log(response.data)
        redirectOnSuccess(response.data.meetup.uri)
    } catch(e){
        console.log(e)
    }
} 

export const editMeetup = (formProps, uri, redirectOnSuccess) => async dispatch => {
    console.log('edit meetup')
    var date = new Date(formProps.date)
    var params = {...formProps, 
        date: date.getFullYear()+ "-" + ("0"+(date.getMonth()+1)).slice(-2) + "-" + ("0" + date.getDate()).slice(-2)
    }
    try {
        const response = await axiosClient.patch(
            `/api/meetups/${uri}/`, params, {headers: {
                "Authorization": `Bearer ${localStorage.getItem('token')}`
        }})
        dispatch({type: EDIT_MEETUP, payload: response.data})
        console.log(response.data)
        redirectOnSuccess(uri)
    } catch(e){
        console.log(e)
    }
}

export const deleteMeetup = (uri) => async dispatch => {
    try {
        const response = await axiosClient.delete(
            `/api/meetups/${uri}/`, {headers: {
                "Authorization": `Bearer ${localStorage.getItem('token')}`
        }})
        console.log(response)
        dispatch({type: DELETE_MEETUP, payload: uri})
        history.push(`/meetups`)
    } catch(e) {
        console.log(e)
    }
}

export const getMeetupEvents = (uri) => async dispatch => {
    try {
        const response = await axiosClient.get(
            `/api/meetups/${uri}/events/`, {headers: {
                "Authorization": `Bearer ${localStorage.getItem('token')}`
        }})
        console.log(response)
        dispatch({type: GET_MEETUP_EVENTS, payload: {uri: uri, data: response.data}})
    } catch(e) {
        console.log(e)
    }
}

export const addMeetupEvent = (event) => async dispatch => {
    dispatch({type: ADD_MEETUP_EVENT, payload: event.message})
}

export const deleteMeetupEvent = (event) => async dispatch => {
    dispatch({type: DELETE_MEETUP_EVENT, payload: event.message})
}

export const reloadMeetupEvent = (event) => async dispatch => {
    dispatch({type: EDIT_MEETUP_EVENT, payload: event.message})
}

export const voteMeetupEvent = (event) => async dispatch => {
    dispatch({type: VOTE_MEETUP_EVENT, payload: event.message})
}
    
export const decideMeetupEvent = (event) => async dispatch => {
    dispatch({type: EDIT_MEETUP_EVENT, payload: event.message})
}

export const addMeetupMember = (event) => async dispatch => {
    dispatch({type: ADD_MEETUP_MEMBER, payload: event.message})
}

export const addEventOption = (event) => async dispatch => {
    dispatch({type: ADD_EVENT_OPTION, payload: event.message})
}

export const sendMeetupEmails = (uri) => async dispatch => {

    try {
        await axiosClient.post(
            `/api/meetups/${uri}/email/`, {}, {headers: {
                "Authorization": `Bearer ${localStorage.getItem('token')}`
        }})
        dispatch({type: ADD_GLOBAL_MESSAGE, payload: {type: "success", message: "Successfully sent emails"}})
    } catch(e) {
        dispatch({type: ADD_GLOBAL_MESSAGE, payload: {type: "error", message: "Unable to send emails"}})
    }
}