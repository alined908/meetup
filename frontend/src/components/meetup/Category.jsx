import React from 'react'
import {Button} from "@material-ui/core"

const Category = (props) => {
    const [variant, setVariant] = React.useState(props.clicked)

    const handleClick = () => {
        setVariant(variant === 'outlined' ? 'contained' : 'outlined')
    }

    return (
        <Button variant={variant} color="primary" onClick={() => handleClick()}>{props.category.label}</Button>
    )
}

export default Category