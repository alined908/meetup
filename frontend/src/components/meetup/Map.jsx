import React, {Component} from "react"
import ReactMapGL, {Marker, NavigationControl, FlyToInterpolator} from "react-map-gl"
import LocationOnIcon from '@material-ui/icons/LocationOn';
import * as d3 from 'd3-ease'

const token = "pk.eyJ1IjoiZGFuaWVsOTA4OTA4IiwiYSI6ImNrNnR6aTdtbTA0bmwzbW1ya3FncjdlcGkifQ.cBAvCpS5zDNwzyb_GUZaoA"

class Map extends Component {

    constructor(props) {
        super(props);
        this.state = {
          viewport: {
            latitude: this.props.location[0],
            longitude: this.props.location[1],
            zoom: 12.5,
            bearing: 0,
            pitch: 0
          }
        };
    }

    onLoad = () => {
        const viewport = {
            ...this.state.viewport,
            zoom: 15,
            transitionDuration: 6000,
            transitionInterpolator: new FlyToInterpolator(),
            transitionEasing: d3.easeSin
        }
        this.setState({viewport})
    }


    render() {
        return (
          <ReactMapGL
            {...this.state.viewport}
            width="450px"
            height="350px"
            mapStyle="mapbox://styles/mapbox/light-v9"
            onViewportChange={viewport => this.setState({viewport})}
            onLoad={this.onLoad}
            mapboxApiAccessToken={token}
          >
            <div style={{position: 'absolute', right: 25, top: 10}}>
                <NavigationControl showCompass={false}/>
            </div>
            <Marker latitude={this.props.location[0]} longitude={this.props.location[1]}>
                <LocationOnIcon fontSize="large" color="secondary"></LocationOnIcon>
            </Marker>
          </ReactMapGL>
        );
      }
}

export default Map