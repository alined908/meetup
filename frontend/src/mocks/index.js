import moment from 'moment'

//Category
export const category = {
    id: 1, 
    image: null,
    label: "Dessert", 
    api_label: "desserts"
}

//User
export const user = {
    id: 1, 
    email: "daniel@gmail.com", 
    first_name: "Daniel", 
    last_name: "Lee",
    avatar: null
}
export const user2 = {
    id: 2, 
    email: "example2@gmail.com", 
    first_name: "Bob", 
    last_name: "Jim", 
    avatar: null
}
export const user3 = {
    id: 3, 
    email: "example3@gmail.com", 
    first_name: "Bob", 
    last_name: "Jim", 
    avatar: null
}
export const settings = {
    radius: 25, 
    location: null, 
    latitude: null,
    longitude: null
}

export const userWithSettings = {
    ...user,
    settings: {...settings}
}
export const friend = {
    id: 1, 
    user: user2, 
    created_at: moment().format(), 
    chat_room: "something"
}
export const friends = [friend]
export const preference = {
    id: 1, 
    user, 
    category, 
    name: "Dessert", 
    ranking: 1, 
    timestamp: moment().format()
}
export const preferences = [preference]
export const friendInvite = {
    id: 1, 
    timestamp: moment().format(), 
    status: 1, 
    uri: "random", 
    sender: user, 
    receiver: user2
}

//Chat
export const message = {
    id: 1, 
    message: "hello", 
    timestamp: moment().format(), 
    room_id: 1, 
    sender: user, 
    is_notif: false
}
export const message2 = {
    id: 2, 
    message: "hello", 
    timestamp: moment().format(), 
    room_id: 1, 
    sender: user, 
    is_notif: true
}
export const messages = [message, message2]
export const members = {1: user, 2: user2}
export const room = {
    uri: {
        id: 1, 
        uri: "abc", 
        name: "Room", 
        timestamp: moment().format(), 
        members: members, 
        friendship: 1, 
        meetup:null, 
        notifs: 0
    }
}
export const rooms = {
    friend: {
        id: 1, 
        uri: "abc", 
        name: "Room", 
        timestamp: moment().format(), 
        members: members, 
        friendship: 1, 
        meetup:null, 
        notifs: 0
    },
    meetup: {
        id: 2, 
        uri: "xyz", 
        name: "Room", 
        timestamp: moment().format(), 
        members: members, 
        friendship: null, 
        meetup:1, 
        notifs: 3
    }
}

//General
export const error = {
    message: "Something went wrong."
}

//Meetup
export const meetupMember1 = {
    1: {
        id: 1,
        user,
        ban: false,
        admin: true
    }
}

export const meetupMember2 = {
    2: {
        id: 2,
        user: user2,
        ban: false,
        admin: false
    }
}

export const meetupMember3 = {
    3: {
        id: 3,
        user: user3,
        ban: false,
        admin: false
    }
}

export const meetupMembers = {...meetupMember1, ...meetupMember2}

export const meetup = {
    uri: {
        id: 1, 
        name: "Meetup", 
        uri: "uri", 
        creator: user, 
        location: "location", 
        date: moment().format("YYYY-MM-DD"),
        members: meetupMembers,
        notifs: 0,
        public: true,
        categories: [category],
        latitude: 34.228754, 
        longitude: -118.2351192,
        events: {},
        isMeetupEventsInitialized : true
    }
}

export const meetups = {...meetup}

export const event = {
    1: {
        id: 1,
        meetup: 1,
        creator: user,
        title: "Dessert",
        start: moment().format(),
        end: moment().format(),
        chosen: null,
        categories: [category],
        options: {},
        price: "1, 2",
        distance: 16000,
        entries: {},
        random: true
    }
}
export const events = {...event}

export const option = {
    1 : {
        id: 1,
        score: 0,
        option: null,
        restaurant: {},
        votes: {},
        banned: false
    }
}

export const options = {...option}

export const tokens = {
    access: "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ0b2tlbl90eXBlIjoiYWNjZXNzIiwiZXhwIjoxNTg4MjI5OTEzLCJqdGkiOiIwOTg2YTg0MDA1ZGM0OGQ1ODEzYmIyNzE1OWViNGUyZCIsInVzZXJfaWQiOjEsInVzZXIiOnsiaWQiOjEsImVtYWlsIjoiZGFuaWVsQGdtYWlsLmNvbSIsImZpcnN0X25hbWUiOiJEYW5pZWwiLCJsYXN0X25hbWUiOiJMZWUiLCJhdmF0YXIiOm51bGwsInNldHRpbmdzIjp7InJhZGl1cyI6MjUsImxvY2F0aW9uIjpudWxsLCJsYXRpdHVkZSI6bnVsbCwibG9uZ2l0dWRlIjpudWxsfX19.FbffRthOKX5F0TGvqA5mZXuXUWNkUojzKexDGCoP518",
    refresh : "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ0b2tlbl90eXBlIjoicmVmcmVzaCIsImV4cCI6MTU4ODM4Mjk0NiwianRpIjoiN2ZhNGY2MDE4OWM3NDA5ZmFjOTMwYmE4ZTRlMWFhMTMiLCJ1c2VyX2lkIjoxLCJ1c2VyIjp7ImlkIjoxLCJlbWFpbCI6ImRhbmllbEBnbWFpbC5jb20iLCJmaXJzdF9uYW1lIjoiRGFuaWVsIiwibGFzdF9uYW1lIjoiTGVlIiwiYXZhdGFyIjpudWxsLCJzZXR0aW5ncyI6eyJyYWRpdXMiOjI1LCJsb2NhdGlvbiI6bnVsbCwibGF0aXR1ZGUiOm51bGwsImxvbmdpdHVkZSI6bnVsbH19fQ.qqaY4Ye7rmyOKLbDS2EKz4SGWH0fwZ0MvrGP6NJ3IRE"
}