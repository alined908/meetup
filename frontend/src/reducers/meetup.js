import {GET_MEETUPS, ADD_MEETUP, DELETE_MEETUP, CLEAR_STORE, VOTE_MEETUP_EVENT, ADD_MEETUP_EVENT, GET_MEETUP_EVENTS, DELETE_MEETUP_EVENT, EDIT_MEETUP_EVENT} from '../constants/action-types';

const defaultState = {
    meetups: {},
    activeMeetup: null,
    isMeetupsInitialized: false,
    isMeetupEventsInitialized: false,
}

export default function meetupReducer(state = defaultState, action){
    switch (action.type){
        case GET_MEETUPS:
            return {...state, meetups: action.payload, isMeetupsInitialized: true}
        case ADD_MEETUP:
            return {...state, meetups: {...state.meetups, [action.payload.uri]: action.payload}}
        case DELETE_MEETUP:
            var meetups = {}
            Object.keys(state.meetups).map((key) => {
                if (key !== action.payload){
                    meetups[key] = state.meetups[key]
                }
            })
            return {...state, meetups: meetups}
        case GET_MEETUP_EVENTS:
           var meetups = {}
           const data = action.payload.data
           Object.keys(state.meetups).map((key) => 
               (key === action.payload.uri) ? meetups[key] = ({...state.meetups[key], events: data}) : (meetups[key] = state.meetups[key])
           )
            return {...state, meetups: meetups, isMeetupEventsInitialized: true}
        case ADD_MEETUP_EVENT:
            return {...state, 
                        meetups: {...state.meetups, 
                            [action.payload.meetup]: {...state.meetups[action.payload.meetup], 
                                events: {...state.meetups[action.payload.meetup].events, 
                                    ...action.payload.event}}}}
        case EDIT_MEETUP_EVENT:
            const event_id = action.payload.event_id
            return {...state, 
                        meetups: {...state.meetups, 
                            [action.payload.meetup]: {...state.meetups[action.payload.meetup],
                                events: {...state.meetups[action.payload.meetup].events, [event_id]: action.payload.event}}}}
        case VOTE_MEETUP_EVENT:
            var meetup = action.payload.meetup
            return {...state, meetups: {...state.meetups, 
                            [meetup]: {...state.meetups[meetup], 
                                events: {...state.meetups[meetup].events, 
                                    [action.payload.event]: {...state.meetups[meetup].events[action.payload.event], 
                                        options: {...state.meetups[meetup].events[action.payload.event].options, 
                                            [action.payload.option_id]: action.payload.option
                    }}}}}}
        case DELETE_MEETUP_EVENT:
            var events = {}
            const uri = action.payload.uri
            Object.keys(state.meetups[uri].events).map((key) => {
                if (key !== action.payload.event.toString()){
                    events[key] = state.meetups[uri].events[key]
                }
            })
            return {...state, meetups: {...state.meetups, [uri] : {...state.meetups[uri], events: events}}}
        case CLEAR_STORE:
            return defaultState
        default:
            return state
    }
}