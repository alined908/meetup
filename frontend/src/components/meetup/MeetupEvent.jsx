import React, {Component} from 'react'
import {Button} from "@material-ui/core"
import {connect} from 'react-redux'
import {deleteMeetupEvent} from "../../actions/meetup"
import moment from "moment"
import {Cached as CachedIcon, Edit as EditIcon, Schedule as ScheduleIcon, Delete as DeleteIcon, Error as ErrorIcon} from '@material-ui/icons'
import {IconButton, Typography, Grid, Grow, Tooltip, Avatar} from '@material-ui/core'
import {compose} from 'redux';
import {Restauraunt, Map} from '../components'
import {Link} from 'react-router-dom'

class MeetupEvent extends Component {
    handleEdit = () => {
        console.log("handle edit function called")
    }

    handleDelete = () => {
        if (window.confirm("Are you sure you want to delete")){
            this.props.socket.deleteMeetupEvent({uri: this.props.uri, event: this.props.event.id})
        }
    }

    handleReload = () => {
        this.props.socket.reloadMeetupEvent({meetup: this.props.uri, event: this.props.event.id})
    }

    handleDecide = () => {
        this.props.socket.decideMeetupEvent({meetup: this.props.uri, event: this.props.event.id, random: false})
    }

    handleRandom = () => {
        this.props.socket.decideMeetupEvent({meetup: this.props.uri, event: this.props.event.id, random: true})
    }

    handleRedecide = () => {
        this.props.socket.redecideMeetupEvent({meetup: this.props.uri, event:this.props.event.id})
    }

    handlePriceChips = (prices) => {
        var priceList = prices.replace(/\s/g, '').split(",");
        for (var i = 0; i < priceList.length; i++){
            priceList[i] = "$".repeat(parseInt(priceList[i]))
        }
        return priceList;
    }

    render () {
        const event = this.props.event

        const renderHeader = (number) => {
            return (
                <div className="meetup-event-header elevate">
                    <div className="inner-header smaller-header">
                        <Typography variant="h5">#{number+1} - {event.title}</Typography>
                        <div className="inner-header-middle">
                            <div className="inner-header-icons"><ScheduleIcon/> {moment(event.start).local().format("h:mm A")} - {moment(event.end).local().format("h:mm A")}</div>
                        </div>
                        <div>
                            {this.props.isUserMember && renderActions()}
                        </div>
                    </div>
                </div>
            )
        }

        const renderFourSquare = (options) => {
            const keys = Object.keys(options)
            
            return (
                <div className="foursquare">
                    {keys.length > 1 ?
                        <Grid justify="center" container spacing={3}>
                            {keys.map((key, index) => 
                                <Grow in={true} timeout={300 * (index + 1)}>
                                    <Grid item justify={index % 2 === 0 ? "flex-end" : "flex-start"} container xs={12} md={6} sm={12}>
                                        <Restauraunt socket={this.props.socket} key={key} full={true} isUserMember={this.props.isUserMember} event={this.props.event.id} meetup={this.props.uri} data={options[key]}/>
                                    </Grid>
                                </Grow>
                            )}
                        </Grid>:
                        <div className="explanation elevate">
                            <span style={{marginRight: "1rem"}}><ErrorIcon style={{color: "rgb(255, 212, 96)"}}/></span>
                            <span>No options available. 
                            This may be due to no options being available at this time or the categories specified don't have enough options.
                            Press the edit button to add more categories or change time if inputted incorrectly.</span>
                        </div>
                    }
                </div>
            )
        }

        const renderActions = () => {
            return (
                <div className="mte-actions">
                    {!this.props.chosen && 
                        <Tooltip title="Reload">
                            <IconButton onClick={() => this.handleReload()} color="primary" aria-label="reload">
                                <CachedIcon />
                            </IconButton>
                        </Tooltip>
                    }
                    <Link to={`/meetups/${this.props.uri}/events/${this.props.event.id}/edit`}>
                        <Tooltip title="Edit">
                            <IconButton color="inherit" aria-label="edit">
                                <EditIcon></EditIcon>
                            </IconButton>
                        </Tooltip>
                    </Link>
                    <Tooltip title="Delete">
                        <IconButton onClick={() => this.handleDelete()} color="secondary" aria-label="delete">
                            <DeleteIcon />
                        </IconButton>
                    </Tooltip>
                </div>
            )
        }

        const renderFinalizeActions = () => {
            return (
                <div className="mte-factions">
                    {!this.props.chosen && <Button className="button rainbow" size="small" variant="contained" color="primary" onClick={() => this.handleDecide()}>Decide</Button>}
                    {!this.props.chosen && <Button className="button rainbow" size="small" variant="contained" color="primary" onClick={() => this.handleRandom()}>Random</Button>}
                    {this.props.chosen && <Button className="button rainbow" size="small" variant="contained" color="primary" onClick={() => this.handleRedecide()}>Redecide</Button>}
                </div>
            )
        }

        const renderChosen = (chosen) => {
            const option = JSON.parse(chosen.option)
            const position = [option.coordinates.latitude, option.coordinates.longitude]

            return (
                <div className="chosen">
                    <div className="chosen-restauraunt elevate">
                        <Restauraunt socket={this.props.socket} isUserMember={this.props.isUserMember} key={chosen.id} full={false} event={this.props.event.id} meetup={this.props.uri} data={chosen}></Restauraunt>
                    </div>
                    <div className="chosen-map elevate2">
                        <div className="map-wrapper">
                            <Map location={position}/>
                        </div>
                    </div>
                </div>
            )
        }

        return (
            <div className="meetup-event">
                {renderHeader(this.props.number)}
                <div className="second-header smaller-header elevate">
                    <div className="second-header-left">
                        <Typography variant="h6">Categories </Typography>
                        {event.categories.map((category) => 
                            <Link to={`/category/${category.api_label}`}>
                                <span className="category-chip elevate category-chip-hover">
                                    <Avatar style={{width: 20, height: 20}} src={`${process.env.REACT_APP_S3_STATIC_URL}${category.api_label}.png`} variant="square"/>
                                    {category.label}
                                </span>
                            </Link>
                        )}
                    </div>
                    <div className="second-header-left">
                        <Typography variant="h6">Price </Typography>  
                        {(this.handlePriceChips(event.price)).map((price) => 
                            <span className="category-chip elevate">{price}</span>
                        )
                        }
                    </div>
                    <div className="second-header-left">
                        <Typography variant="h6">Distance </Typography>
                        <span className="category-chip elevate">
                            {(Math.round(event.distance * 0.000621371192)).toFixed(2)}
                        </span>
                    </div>
                    {this.props.isUserMember && renderFinalizeActions()}
                </div>
                {!this.props.chosen && renderFourSquare(event.options)}
                {this.props.chosen && renderChosen(event.options[this.props.chosen])}
                
            </div>
        )
    }
}

function mapStateToProps(state, props) {
    return {
        chosen: state.meetup.meetups[props.uri].events[props.event.id].chosen
    }
}

const mapDispatchToProps = {
    deleteMeetupEvent
}

export default compose(
    connect(mapStateToProps, mapDispatchToProps), 
  )(MeetupEvent)