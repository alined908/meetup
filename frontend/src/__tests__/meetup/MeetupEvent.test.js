import React from 'react'
import {shallow} from 'enzyme';
import {UnderlyingMeetupEvent} from "../../components/meetup/MeetupEvent"
import * as mocks from "../../mocks"

const props = {
    user: mocks.user,
    event: mocks.event[1],
    number: 1,
    socket: {
        deleteMeetupEvent: jest.fn(),
        reloadMeetupEvent: jest.fn(),
        decideMeetupEvent: jest.fn(),
        redecideMeetupEvent: jest.fn(),
        addEventOption: jest.fn(),
    },
    uri: mocks.meetup.uri.uri,
    coords: {latitude: 32.22, longitude: 32.33},
    deleteMeetupEvent: jest.fn(),
    addGlobalMessage: jest.fn(),
    chosen: false,
    isUserMember: true
}

describe("MeetupEvent tests general", () => {

    afterEach(() => {
        jest.clearAllMocks();
    });

    it ('renders without crashing', () => {
        const wrapper = shallow(<UnderlyingMeetupEvent {...props}/>)
        expect(wrapper).toMatchSnapshot()
    })

    it('opens form modal onclick', () => {
        const wrapper = shallow(<UnderlyingMeetupEvent {...props}/>)
        expect(wrapper.state('editMeetupEventForm')).toBe(false)
        expect(wrapper.find("MeetupEventForm")).toHaveLength(0)
        const button = wrapper.find('[aria-label="edit"]')
        button.simulate('click')
        expect(wrapper.state('editMeetupEventForm')).toBe(true)
    })

    it('should handle search value click if restaurant doesnt already exist as option', () => {
        const newProps = {...props, event: {...mocks.event[1], options: mocks.options}}
        const wrapper = shallow(<UnderlyingMeetupEvent {...newProps}/>)
        wrapper.setState({searchInput: "food"})
        wrapper.instance().handleSearchValueClick({}, {id: "rst2"})
        expect(props.addGlobalMessage.mock.calls.length).toBe(1)
        expect(props.socket.addEventOption.mock.calls.length).toBe(1)
        expect(wrapper.state("searchInput")).toBe("")
    })

    it('should handle search value click if restaurant already exist as option', () => {
        const newProps = {...props, event: {...mocks.event[1], options: mocks.options}}
        const wrapper = shallow(<UnderlyingMeetupEvent {...newProps}/>)
        wrapper.setState({searchInput: "food"})
        wrapper.instance().handleSearchValueClick({}, {id: "rst"})
        expect(props.addGlobalMessage.mock.calls.length).toBe(1)
        expect(props.socket.addEventOption.mock.calls.length).toBe(0)
        expect(wrapper.state("searchInput")).toBe("")
    })
})

describe("MeetupEvent tests not chosen", () => {
    window.confirm = jest.fn(() => true)

    afterEach(() => {
        jest.clearAllMocks();
    });

    it ('shows reload button and sends action on click', () => {
        const wrapper = shallow(<UnderlyingMeetupEvent {...props}/>)
        const button = wrapper.find('ProgressIcon')
        expect(button).toHaveLength(1)
        wrapper.instance().handleReload()
        expect(props.socket.reloadMeetupEvent.mock.calls.length).toBe(1)
    })

    it('shows add option action and shows searchbar onclick', () => {
        const wrapper = shallow(<UnderlyingMeetupEvent {...props}/>)
        const button = wrapper.find('[aria-label="add-option"]')
        expect(button).toHaveLength(1)
        button.simulate('click')
        expect(wrapper.state("searchOpen")).toBe(true)
        expect(wrapper.find(".addOptionSearch")).toHaveLength(1)
    })

    it ('shows reload button and sends action on click', () => {
        const wrapper = shallow(<UnderlyingMeetupEvent {...props}/>)
        const button = wrapper.find('[aria-label="delete"]')
        expect(button).toHaveLength(1)
        button.simulate('click')
        expect(props.socket.deleteMeetupEvent.mock.calls.length).toBe(1)
    })

    it ('if there are no options doesnt show decide and random and displays helper text', () => {
        const wrapper = shallow(<UnderlyingMeetupEvent {...props}/>)
        const decideButton = wrapper.find('[aria-label="decide"]')
        expect(decideButton).toHaveLength(0)
        const randomButton = wrapper.find('[aria-label="random"]')
        expect(randomButton).toHaveLength(0)
        expect(wrapper.find(".explanation").text()).toContain("No options available.")
    })

    it ('shows decide and random buttons and sends actions on click if options exist', () => {
        const newProps = {...props, event: {...mocks.event[1], options: mocks.options}}
        const wrapper = shallow(<UnderlyingMeetupEvent {...newProps}/>)
        const decideButton = wrapper.find('[aria-label="decide"]')
        expect(decideButton).toHaveLength(1)
        decideButton.simulate('click')
        expect(props.socket.decideMeetupEvent.mock.calls.length).toBe(1)
        const randomButton = wrapper.find('[aria-label="random"]')
        expect(randomButton).toHaveLength(1)
        randomButton.simulate('click')
        expect(props.socket.decideMeetupEvent.mock.calls.length).toBe(2)
    })
})

describe("MeetupEvent tests chosen", () => {
    const newProps = {...props, chosen: 1, event: {...mocks.event[1], options: mocks.options}}

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render option and map', () => {
        const wrapper = shallow(<UnderlyingMeetupEvent {...newProps}/>)
        expect(wrapper.find("Map")).toHaveLength(1)
        expect(wrapper.find("Connect(MeetupEventOption)")).toHaveLength(1)
    })

    it ('should show redecide button and send event onclick', () => {
        const wrapper = shallow(<UnderlyingMeetupEvent {...newProps}/>)
        const redecideButton = wrapper.find('[aria-label="redecide"]')
        expect(redecideButton).toHaveLength(1)
        redecideButton.simulate('click')
        expect(props.socket.redecideMeetupEvent.mock.calls.length).toBe(1)
    })
})