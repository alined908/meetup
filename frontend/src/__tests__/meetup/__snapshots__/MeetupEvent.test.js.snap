// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeetupEvent tests general renders without crashing 1`] = `
<div
  className="event"
  id="event-1"
>
  <WithStyles(ForwardRef(Paper))
    elevation={3}
  >
    <div
      className="header smallerheader"
    >
      <WithStyles(ForwardRef(Typography))
        variant="h5"
      >
        #
        2
         - 
        Dessert
      </WithStyles(ForwardRef(Typography))>
      <div
        className="headerInfo"
      >
        <div
          className="headerIcons"
        >
          <ScheduleIcon />
          6:07 AM
           - 
          6:07 AM
        </div>
        <div
          className="headerIcons"
        >
          <img
            alt="&#9787;"
            src="https://meetup-static.s3-us-west-1.amazonaws.com/static/general/panda.png"
            style={
              Object {
                "height": 20,
                "marginLeft": 10,
                "width": 20,
              }
            }
          />
          - 
          Daniel
           
          Lee
        </div>
      </div>
      <div>
        <div
          className="eventActions"
        >
          <WithStyles(ForwardRef(Tooltip))
            title="Add Option"
          >
            <WithStyles(ForwardRef(IconButton))
              aria-label="add-option"
              onClick={[Function]}
              style={
                Object {
                  "color": "#4caf50",
                }
              }
            >
              <SearchIcon />
            </WithStyles(ForwardRef(IconButton))>
          </WithStyles(ForwardRef(Tooltip))>
          <WithStyles(ForwardRef(Tooltip))
            title="Reload"
          >
            <div
              style={
                Object {
                  "minHeight": 48,
                  "width": 48,
                }
              }
            >
              <ProgressIcon
                ariaLabel="reload"
                check={false}
                color="primary"
                disabled={false}
                handleClick={[Function]}
                icon={<UNDEFINED />}
              />
            </div>
          </WithStyles(ForwardRef(Tooltip))>
          <WithStyles(ForwardRef(Tooltip))
            title="Edit"
          >
            <WithStyles(ForwardRef(IconButton))
              aria-label="edit"
              color="inherit"
              onClick={[Function]}
            >
              <EditIcon />
            </WithStyles(ForwardRef(IconButton))>
          </WithStyles(ForwardRef(Tooltip))>
          <WithStyles(ForwardRef(Tooltip))
            title="Delete"
          >
            <WithStyles(ForwardRef(IconButton))
              aria-label="delete"
              color="secondary"
              onClick={[Function]}
            >
              <DeleteIcon />
            </WithStyles(ForwardRef(IconButton))>
          </WithStyles(ForwardRef(Tooltip))>
        </div>
      </div>
    </div>
  </WithStyles(ForwardRef(Paper))>
  <WithStyles(ForwardRef(Paper))
    className="smallerHeader secondHeader"
    elevation={3}
  >
    <div
      className="secondHeaderLeft"
    >
      <WithStyles(ForwardRef(Typography))
        variant="h6"
      >
        Categories 
      </WithStyles(ForwardRef(Typography))>
      <Link
        key="1"
        to="/category/desserts"
      >
        <span
          className="categoryChip categoryChipHover"
        >
          <WithStyles(ForwardRef(Avatar))
            src="undefineddesserts.png"
            style={
              Object {
                "height": 20,
                "width": 20,
              }
            }
            variant="square"
          />
          Dessert
        </span>
      </Link>
    </div>
    <div
      className="secondHeaderLeft"
    >
      <WithStyles(ForwardRef(Typography))
        variant="h6"
      >
        Price 
      </WithStyles(ForwardRef(Typography))>
      <span
        className="categoryChip"
        key="0"
      >
        $
      </span>
      <span
        className="categoryChip"
        key="1"
      >
        $$
      </span>
    </div>
    <div
      className="secondHeaderLeft"
    >
      <WithStyles(ForwardRef(Typography))
        variant="h6"
      >
        Distance 
      </WithStyles(ForwardRef(Typography))>
      <span
        className="categoryChip"
      >
        10.00
      </span>
    </div>
    <div
      className="eventDecideActions"
    />
  </WithStyles(ForwardRef(Paper))>
  <div
    className="foursquare"
  >
    <WithStyles(ForwardRef(Paper))
      className="explanation"
      elevation={3}
    >
      <span
        style={
          Object {
            "marginRight": "1rem",
          }
        }
      >
        <ErrorIcon
          style={
            Object {
              "color": "rgb(255, 212, 96)",
            }
          }
        />
      </span>
      <span>
        No options available. This may be due to no options being available at this time or the categories specified don't have enough options. Press the edit button to add more categories or change time if inputted incorrectly.
      </span>
    </WithStyles(ForwardRef(Paper))>
  </div>
</div>
`;
